### PR TITLE
Configure git authentication by http url

### PR DESCRIPTION
**Pull request by yufeih**
This PR enables per domain/baseURL git authentication configuration using `git -c http.{baseURL}.extraheader`. This unifies the way we handle git authentication and blob authentication.

http configuration has been changed to a simple mapping to align with git.

```yaml
http:
  {baseURL}:
    query: {query}
    headers:
       {key}: {value}
```

For GitHub, here is an example config:
```yaml
http:
  https://github.com:
    headers:
      Authorization: basic {base_64_encoded_github_personal_access_token}
```

@fenxu @superyyrrzz

### CODE REVIEW

**Comment by OsmondJiang** [comment]
>                 syntax.DefineOption("github-token", ref options.GitHubToken, "The GitHub token used to get contribution information from GitHub API");

also remove this?

---
Refers to: src/docfx/cli/Program.cs:98 in 5cd8d46. [](commit_id = 5cd8d46d53c6c246be99341274963b1ed2601b82, deletion_comment = False)

File: src/docfx/lib/HttpClientUtility.cs [hunk 1]
```diff
@@ -39,21 +39,21 @@ public static async Task<HttpResponseMessage> GetAsync(string requestUri, HttpSe
...
-        private static HttpRequestMessage CreateHttpRequestMessage(string requestUri, HttpSecretConfig[] rules)
+        private static HttpRequestMessage CreateHttpRequestMessage(string requestUri, Config config)
         {
             var message = new HttpRequestMessage();
 
-            foreach (var rule in rules)
+            foreach (var (baseUrl, rule) in config.Http)
             {
-                if (rule.Match(requestUri))
+                if (requestUri.StartsWith(baseUrl))
```
Comment (OsmondJiang) [inline src/docfx/lib/HttpClientUtility.cs]: >StartsWith [](start = 31, length = 10)

case sensitive?

File: src/docfx/lib/HttpClientUtility.cs [hunk 2]
```diff
@@ -39,21 +39,21 @@ public static async Task<HttpResponseMessage> GetAsync(string requestUri, HttpSe
...
                 {
                     // TODO: merge query if requestUri also contains query
                     message.RequestUri = new Uri(requestUri + rule.Query);
```
Comment (OsmondJiang) [inline src/docfx/lib/HttpClientUtility.cs]: > message.RequestUri = new Uri(requestUri + rule.Query); [](start = 19, length = 55)

add a spec test for that?

**Comment by yufeih** [comment]
I'll update the spec soon

File: src/docfx/config/Config.cs [hunk 3]
```diff
@@ -105,18 +105,13 @@ internal sealed class Config
         /// <summary>
         /// Gets the authorization keys for required resources access
         /// </summary>
-        public readonly HttpConfig Http = new HttpConfig();
```
Comment (OsmondJiang) [inline src/docfx/config/Config.cs]: also update the spec [https://github.com/dotnet/docfx/blob/v3/docs/designs/credential.md](https://github.com/dotnet/docfx/blob/v3/docs/designs/credential.md)

File: src/docfx/restore/RestoreGit.cs [hunk 4]
```diff
@@ -97,16 +97,24 @@ private static async Task<List<(string href, string head)>> AddWorkTrees(string
 
             Task FetchOrCloneRepo()
             {
+                var gitConfigs =
+                    from http in config.Http
+                    where url.StartsWith(http.Key)
+                    from header in http.Value.Headers
+                    select $"-c http.{http.Key}.extraheader=\"{header.Key}: {header.Value}\"";
```
Comment (superyyrrzz) [inline src/docfx/restore/RestoreGit.cs]: >{http.Key}. [](start = 37, length = 11)

Is this part needed? Seems git only supports `http.extraHeader` [https://github.com/git/git/blob/a42d7b6a5b5addcce01f483547ad96825c7a96e0/Documentation/config.txt#L1733](https://github.com/git/git/blob/a42d7b6a5b5addcce01f483547ad96825c7a96e0/Documentation/config.txt#L1733)

Comment (yufeih) [reply src/docfx/restore/RestoreGit.cs]: https://github.com/git/git/blob/master/Documentation/config.txt#L2052

File: src/docfx/lib/HttpClientUtility.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/HttpClientUtility.cs]: yes. Url comparison is complex, it is temporarily using case sensitive starts with.

File: test/docfx.Test/e2e/E2ETest.cs [hunk 5]
```diff
@@ -143,11 +148,29 @@ private static async Task<(string docsetPath, E2ESpec spec)> CreateDocset(string
                 Process.Start(new ProcessStartInfo("git", "submodule update --init") { WorkingDirectory = docsetPath }).WaitForExit();
             }
 
+            var replaceEnvironments =
+                spec.Environments.Length > 0 &&
+                !spec.Environments.Any(env => string.IsNullOrEmpty(Environment.GetEnvironmentVariable(env)));
+
+            var skip = spec.Environments.Length > 0 && !replaceEnvironments;
```
Comment (live1206) [inline test/docfx.Test/e2e/E2ETest.cs]: >var skip = spec.Environments.Length > 0 && !replaceEnvironments; [](start = 12, length = 64)

could it just be as below?
```csharp
var skip = spec.Environments.Any(env => string.IsNullOrEmpty(Environment.GetEnvironmentVariable(env)));
```

**Comment by OsmondJiang** [comment]
>                 syntax.DefineOption("github-token", ref options.GitHubToken, "The GitHub token used to get contribution information from GitHub API");

？

---
In reply to: [429220226](https://github.com/dotnet/docfx/pull/3490#issuecomment-429220226) [](ancestors = 429220226)

---
Refers to: src/docfx/cli/Program.cs:98 in 5cd8d46. [](commit_id = 5cd8d46d53c6c246be99341274963b1ed2601b82, deletion_comment = False)

**Pull request merged**